### PR TITLE
feat: on run, build only when filesystem changed

### PR DESCRIPTION
**Pull request by lkingland**
:gift: client detects Function build staleness via `.Built(path)`
:gift: `run` command accepts `auto|true|false` for `--build`

Upon build, a hash of the filesystem at time of build is stored in the Function runtime directory `.func`.  This stamp is exposed via a new `.Built` function on the client.

This function is utilized to implement more intelligent build logic on the `run` command as initial acceptance testing, and will be integrated into the `deploy` command in a subsequent PR if acceptable.

/kind enhancement

```release-notes
On `run`, functions are only built when something on the filesystem changes
```

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -885,6 +910,89 @@ func (c *Client) Push(ctx context.Context, path string) (err error) {
...
+	// It's a pretty good chance the thing doesn't need to be rebuilt, though
+	// of course filesystem racing conditions do exist, including both direct
+	// source code modifications or changes to the image cache.
+	return true
+}
+
+// fingerprint returns a hash of the filenames and modification timestamps of
+// the files within a Function's root.
+func fingerprint(f Function) (string, error) {
+	b := bytes.Buffer{}
```
Comment (matejvasek) [inline client.go:971]: Nit: you might not need buffer.
```go
h := sha256.New()
// ...
fmt.Fprintf(h, "%v:%v:", path, info.ModTime().UnixNano());
s := h.Sum(nil)
```

File: client.go [hunk 2]
```diff
@@ -885,6 +910,89 @@ func (c *Client) Push(ctx context.Context, path string) (err error) {
...
+
+	// Function has a populated image, existing buildstamp, and the calculated
+	// fingerprint has not changed.
+	// It's a pretty good chance the thing doesn't need to be rebuilt, though
+	// of course filesystem racing conditions do exist, including both direct
+	// source code modifications or changes to the image cache.
+	return true
+}
+
+// fingerprint returns a hash of the filenames and modification timestamps of
```
[the remark this answers is not in the document]
Comment (lkingland) [reply client.go:968]: Seems unnecessary to read every file into ram and then into through a buffer into a tar file, when a filename plus last-modified timestamp accomplishes the same goal; especially when considering the user's directory may include arbitrarily large files such as ISOs, or all the images of a webapp.

File: client.go (hunk 1 above)
Comment (lkingland) [reply client.go:971]: That's a good point; changing that...

File: client.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (matejvasek) [reply client.go:968]: It is steamed -- entire file is not loaded into memory, it's done by chunks. but yes its slower, 1G file takes 1.5sec.

[the remark this answers is not in the document]
Comment (matejvasek) [reply client.go:968]: Memory consumption is constant -- under 1M.

[the remark this answers is not in the document]
Comment (lkingland) [reply client.go:968]: I liked your suggestion to write directly to the sum buffer.  But as for your question "Why not content hashing?" I would say that in general, using the least amount of resources and data to accomplish a specific task is a worthy goal.  I like your suggested implementation, of course!  I just think it's not necessary complexity.

We want to rebuild when the filesystem has changed.  That needs two pieces of data: list of files and lat modified date.   What benefit do we get using an implementation that is slower, uses more memory, is harder on the disk, has several additional dependencies and many more lines of cde to maintain, when this
```
  h := sha256.New()
  err := filepath.Walk(f.Root, func(path string, info fs.FileInfo, err error) error {
    if err != nil {
      return err
    }
    fmt.Fprintf(h, "%v:%v:", path, info.ModTime().UnixNano())
    return nil
  })
  return fmt.Sprintf("%x", h.Sum(nil)), err
```
seems to do just fine?
Perhaps I am not thinking of an edge case here where content of the files is important!

[the remark this answers is not in the document]
Comment (matejvasek) [reply client.go:968]: >Unless there is a compelling reason content hashing is more appropriate, I'll side with simplicity.

Timestamps are all right for now I guess.
Only only usecase I can think of is when users unintentionally changes modtime modify-revert-save in IDE.

**Review by zroubalik — APPROVED** [review]
/lgtm

/hold for others

**Review by lance — COMMENTED** [review]
/lgtm
this is a nice addition - thanks!

File: client.go [hunk 3]
```diff
@@ -859,7 +893,7 @@ func (c *Client) Push(ctx context.Context, path string) (err error) {
 		return
 	}
 
-	if !f.Built() {
+	if !f.HasImage() {
```
[the remark this answers is not in the document]
Comment (lkingland) [reply client.go:897]: That might be a good idea.  Adding this to my punts list.

**Comment by zroubalik** [comment]
@lkingland could you please fix the typo?

**Comment by lkingland** [comment]
> @lkingland could you please fix the typo?

Yes, fixed in the next PR (I don't want to run the springboot E2E test lottery again and it's an optional test)

**Pull request merged**